File: Competitive Programming Solutions/WC/wc18c3s2.java
import java.util.*;
import java.io.*;
public class GymTour{
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static int cnt = 0, cnt2 = 0, max = 0, F;
	static List<Integer>[] adj;
	static boolean[] gym;
	static void dfs1(int n, int pre, int depth) {
		for(int i : adj[n]) {
			if(i != pre) { 
				dfs1(i, n, depth+1);
				if(gym[i]) { 
					cnt++;
					gym[n]=true; 
				}
			}
		}
		if(gym[n])max = Math.max(max, depth);
	}
	static void dfs2(int n, int pre, int depth) {
		for(int i : adj[n]) {
			if(i != pre) { 
				dfs2(i, n, depth+1);
				if(gym[i]||i==F) { 
					cnt2++;
					gym[n]=true; 
				}
			}
		}
	}
	public static void main(String[] args) throws IOException{
		int N = readInt(), K = readInt();
		F = readInt();
		gym = new boolean[N+1];
		for(int i = 0; i<K; i++)gym[readInt()]=true;
		adj = new ArrayList[N+1];
		for(int i = 1; i<=N; i++)adj[i]=new ArrayList<Integer>();
		for(int i = 1; i<N; i++) {
			int a = readInt(), b = readInt();
			adj[a].add(b);
			adj[b].add(a);
		}
		dfs1(1, -1, 0);  
		dfs2(1, -1, 0);
		System.out.println(Math.min(2*cnt-max, cnt2));
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}

File: Contest Problems/WC/wc18c3s2.cpp
#include <bits/stdc++.h>
#include <iostream>
using namespace std;
int N,K,F,cnt = 0, cnt2 = 0, m = 0;
vector<int> adj[1000001];
bool gym[1000001];
void dfs1(int n, int pre, int depth) {
	for(int i : adj[n]) {
		if(i != pre) { 
			dfs1(i, n, depth+1);
			if(gym[i]) { 
				cnt++;
				gym[n]=true; 
			}
		}
	}
	if(gym[n])m=max(m,depth);
}
void dfs2(int n, int pre, int depth) {
	for(int i : adj[n]) {
		if(i != pre) { 
			dfs2(i, n, depth+1);
			if(gym[i]||i==F) { 
				cnt2++;
				gym[n]=true; 
			}
		}
	}
}
int main() {
    cin.sync_with_stdio(0);
    cin.tie(0);
    cin>>N>>K>>F;
	for(int i = 0,a; i<K; i++){
        cin>>a;
        gym[a]=true;
    }
	for(int i = 1,a,b; i<N; i++) {
		cin>>a>>b;
		adj[a].push_back(b);
		adj[b].push_back(a);
	}
		dfs1(1, -1, 0);  
		dfs2(1, -1, 0);
		cout<<min(2*cnt-m, cnt2)<<"\n";
}